src/types.rs

```diff
@@ -330,3 +331,119 @@ impl DestinationSlot {
         self.dest.get()
     }
 }
+
+/// Helper structure for parsing the result of a multishot [`opcode::RecvMsg`](crate::opcode::RecvMsg).
+#[derive(Debug)]
+pub struct RecvMsgOut<'buf> {
+    buf: &'buf [u8],
+    header: sys::io_uring_recvmsg_out,
+    msghdr_name_len: u32,
+    msghdr_control_len: u32,
+}
+
+impl<'buf> RecvMsgOut<'buf> {
+    const DATA_START: usize = std::mem::size_of::<sys::io_uring_recvmsg_out>();
+
+    /// Parse the data buffered upon completion of a `RecvMsg` multishot operation.
+    ///
+    /// `buffer` is the whole buffer previously provided to the ring, while `name_len`
+    /// and `control_len` are equal to the input lengths provided through `msghdr` in
+    /// the corresponding SQE.
+    pub fn parse(buf: &'buf [u8], name_len: u32, control_len: u32) -> Result<Self, ()> {
+        if buf.len() < std::mem::size_of::<sys::io_uring_recvmsg_out>() {
+            return Err(());
+        }
+
+        let header = sys::io_uring_recvmsg_out {
+            namelen: u32::from_ne_bytes(buf[0..4].try_into().unwrap()),
+            controllen: u32::from_ne_bytes(buf[4..8].try_into().unwrap()),
+            payloadlen: u32::from_ne_bytes(buf[8..12].try_into().unwrap()),
+            flags: u32::from_ne_bytes(buf[12..16].try_into().unwrap()),
+        };
+
+        let total_len = (name_len + control_len + header.payloadlen) as usize;
+        if Self::DATA_START + total_len > buf.len() {
+            return Err(());
+        }
+
+        Ok(Self {
+            buf,
+            header,
+            msghdr_name_len: name_len,
+            msghdr_control_len: control_len,
+        })
+    }
+
+    /// Return the length of the incoming `name` data.
+    ///
+    /// This may be larger than the size of the content returned by
+    /// `name_data()`, if the kernel could not fit all the incoming
+    /// data in the provided buffer size. In that case, name data in
+    /// the result buffer gets truncated.
+    pub fn incoming_name_len(&self) -> u32 {
+        self.header.namelen
+    }
+
+    /// Return whether the incoming name data was larger than the provided limit/buffer.
+    ///
+    /// When `true`, data returned by `name_data()` is truncated and
+    /// incomplete.
+    pub fn is_name_data_truncated(&self) -> bool {
+        self.header.namelen > self.msghdr_name_len
+    }
+
+    /// Message control data, with the same semantics as `msghdr.msg_control`.
+    pub fn name_data(&self) -> &[u8] {
+        let name_start = Self::DATA_START;
+        let name_size = u32::min(self.header.namelen, self.msghdr_name_len);
+        let name_end = name_start + name_size as usize;
+        &self.buf[name_start..name_end]
```

Comment:
Could you add some more details about what you are suggesting here?
My feeling is that what you want to do goes against the limit of having self-references within `RecvMsgOut` (the buffer, plus the subfields pointing within it), but maybe I misunderstood what you have in mind.

Reply:
On the topic of performances, I did tweak the logic a bit so that all accessor methods can use unchecked/saturating math.

Reply:
It is an immutable reference, there is no self-referencing problem. We also don't need to keep original buf, we can use `split_at` to split.

Reply:
Ack, thanks, initially I didn't understand what you were suggesting. I've changed the parsing logic so that the buffer is split upfront and the structure only keeps references to the three specific data slices.